crates/pairwise_nostd_internal/src/bins.rs

```diff
@@ -0,0 +1,193 @@
+//! Implements types to represent "bin edges", used for Histogram buckets and for
+//! distance binning of accumulators.  The [`BinEdges`] trait provides a common
+//! interface that is implemented by [`RegularBinEdges`] and [`IrregularBinEdges`]
+
+/// Super simple. This can be expanded as needed.
+pub trait BinEdges {
+    /// Calculate the bin index for a given value. Values which are equal to
+    /// boundary values are considered part of the higher bin, i.e. intervals
+    /// do not include the right edge.
+    fn bin_index(&self, value: f64) -> Option<usize>;
+
+    fn n_bins(&self) -> usize;
+
+    fn is_empty(&self) -> bool {
+        self.n_bins() == 0
+    }
```

Comment:
Sorry, one last thought. Did you just add this method simply to satisfy clippy's lint about having a `len()` method without an `is_empty()` method?

If so, can we remove this method now that we use replaced `len()` with `n_bins`? By construction I don't think we ever allow `BinEdges` to be empty, right? Or do you think we should keep it?

```suggestion
```